// Copyright 2015 gRPC authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

package gamingstreaming;

// Interface exported by the server.
service GamingStreamingService {
    // A Bidirectional streaming RPC.
    //
    // Accepts a stream of RouteNotes sent while a route is being traversed,
    // while receiving other RouteNotes (e.g. from other users).
    rpc GamingChannel (stream InputCommand) returns (stream FrameSubPacket) {}
}

message FrameSubPacket {
        int32 frame_number = 1;
        int32 packet_number = 2;
        int32 size = 3;
        int64 pts = 4;
        int64 dts = 5;
        int64 flags = 6;
        // field related to frame management
        bytes data = 7;
        // field related to mouse position
        int32 mouse_x = 8; // for mouse
        int32 mouse_y = 9; // for mouse
        bool mouse_is_visible = 10; // for mouse
}

message InputCommand {
        string command = 1;
        int32 event_type = 2; // 0 -> mouse; 1 -> keyboard
        int32 key_code = 3; // for keyboard
        int32 x = 4; // for mouse
        int32 y = 5; // for mouse
        int32 mouse_button = 6; // for mouse
}